In [1]:
# importing the library
import pandas as pd

 În cadrul acestei părți din tutorial o să învățăm cum putem să ne ocupăm de partea de valori care lipsesc din cadrul data frame-ului (missing data) și cum putem și să facem curat prin aceste date (deoarece unele dintre aceste date nu sunt concludente). Aproape fiecare data set cu care lucrăm are în componența lui valori care lipsesc, valori care trebuie curățate sau valori la care dorim să le schimbăm tipul. În cadrul acestei părți din tutorial o să învățăm cum să face toate acestea lucruri. Cu ceea ce învățăm aici o să calculăm pe baza data frame-ului cu studiul de pe stack overflow media de experiență în ani a persoanelor care au participat la acest studiu. 

Pentru început o să vedem cum anume putem să facem drop la valorile care lipsesc. Pentru a crea un data frame cu un set de valori lipsă o să ne utilizăm și de librăria numpy

In [5]:
import pandas as pd
import numpy as np

In [11]:
peoples = {
    'first' : ['Corey', 'Jane', 'John', 'Chris', np.nan, None, 'NA'],
    'last' : ['Schafer', 'Doe', 'Doe', 'Schafer', np.nan, np.nan, 'Missing'],
    'email' : ['CoreySchafer@mail.com', 'JaneDoe@gmail.com', 'JohnDoe@gmail.com', None, np.nan, 'Anonymus@mail.com', 'NA'],
    'age' : ['33', '55', '63', '36', None, None, 'Missing']
}

In [15]:
peoples_df = pd.DataFrame(peoples)

In [16]:
peoples_df

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
3  Chris  Schafer                   None       36
4    NaN      NaN                    NaN     None
5   None      NaN      Anonymus@mail.com     None
6     NA  Missing                     NA  Missing

Putem observa că în cazul acestui data frame avem tot felul de valori care nu ne sunt foarte concludente, precum cele cu 'NaN', 'None', etc. Depinzând de ce anume dorim să facem în cadrul acestui data frame, poate dorim să tratăm diferit anumite cazuri. Un lucru des pe care o să îl facem în cazul datelor lipsă, este să eliminăm aceste date. Pentru situația de față să presupunem că dorim să facem ceva analiză cu aceste date, iar persoanele care nu au nimic setat ca și valoare în oricare dintre coloane nu o să putem face această analiză, prin urmare o să ștergem rândurile care sunt incomplete. Pentru a șterge datele unde nu avem valori o să utilizăm metoda `dropna()`

In [17]:
peoples_df.dropna()

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
6     NA  Missing                     NA  Missing

Când rulăm această coloană fără niciun argument putem vedea că în acest moment avem doar 4 rânduri de date, nu 7 cum era la început. Linia de mai sus a șters datele ce conțineau valori precum 'NaN' sau 'None'. Au mai rămas în cadrul acestui data frame valorile ce le-am creat noi, dar și acestea sun valori de tip missing data (cum ar fi 'NA' sau 'Missing' - o să revenim la acestea imediat)

Metoda 'dropna()' utilizează ceva argumente setate default. O să trecem și noi în cod aceste argumente pentru a înțelege mai bine ce anume face această metodă. Argumentele default care sunt folosite pentru această metodă sunt "axis='index'" și "how='any'". Dacă specificăm aceste argumente cu aceleași valori, atunci metoda trebuie să returneze același rezultat

In [18]:
peoples_df.dropna(axis='index', how='any')

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
6     NA  Missing                     NA  Missing

Argumentul 'axis' este utilizat pentru a specifica de pe ce să șteargă datele când găsește un tip de date NaN. Default este setat la 'index' ceea ce înseamnă că în momentul în care găsește un astfel de missing value în cadrul data frame-ului, atunci o să șteargă întreg rândul. Dacă dorim să ștergem coloana pe care se găsește acea valorea, atunci trebuie să modificăm valorea argumentului la 'columns'. Argumentul 'how' reprezintă criteriile pentru care să șteargă un rând sau o coloană. În modul default este setat la 'any'. Asta înseamnă că se uită la un anumit rând in cadrul data frame-ului și fiind setat la 'any' o să șteargă acel rând dacă găsește în oricare dintre coloane o valoare care lipsește. Poate de multe ori nu dorim să șteargă întregul rând dacă am avea data care lipses de la colona de 'email' sau din alte coloane. Acestui argument putem să îi setăm valoarea 'all'. Astfel, se va șterge rândul doar acolo unde sunt date care lipsesc în fiecare coloană din cadrul acelui rând

In [19]:
peoples_df.dropna(axis='index', how='all')

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
3  Chris  Schafer                   None       36
5   None      NaN      Anonymus@mail.com     None
6     NA  Missing                     NA  Missing

Dacă rulăm acuma acest cod putem observa că au rămas mai multe date în cadrul data frame-ului deoarce s-au păstrat datele unde exista ceva anume în cadrul unor coloane, chiar dacă sunt anumite coloane ce au missing value, s-au șters rândurile unde pentru fiecare coloană din rând era câte un missing value. 

De foarte multe ori dorim să facem ceva analize și pentru asta e nevoie să ștergem anumite date care se găsesc într-o anumită coloană, nu neapărat din toate coloanele din cadrul data frame-ului. Să luăm cazul unde dorim să facem o analiză pentru acest data frame, dar avem neapărat nevoie de adresa de email, restul datlor pot să lipsească. Pentru a specifica o anumită coloană putem utiliza argumentul de `subset`. Acest argument primește ca și valoare o listă în care trecem coloana sau coloanele de unde dorim să ștergem datele de tip missing data

In [21]:
peoples_df.dropna(axis='index', how='any', subset=['email'])

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
5   None      NaN      Anonymus@mail.com     None
6     NA  Missing                     NA  Missing

Putem observa că în acest moment data frame-ul care a rezultat are date în cadrul coloanei de email. De reținut totuși că ultima valoarea, cea de 'NA' este valorea noastră ce a fost introdusă ca și string (o să vedem cum ne putem ocupa de aceste valori). În această situație, argumentul de 'how' nu prea își are rolul deoarece se uită doar în cadrul coloanei de email. Diferența apare în momentul în care se pasează mai multe coloane pentru argumentul 'subset'. Abia atunci utilizarea argumetului 'how' își are rostul, deoarece prin el specificăm dacă dorim să caute valori care lipsesc doar dintr-o coloană specificată la partea de 'subset', sau trebuie să lipsească din toate coloanele.

In [23]:
peoples_df.dropna(axis='index', how='all', subset=['last', 'email'])

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
3  Chris  Schafer                   None       36
5   None      NaN      Anonymus@mail.com     None
6     NA  Missing                     NA  Missing

In [24]:
peoples_df.dropna(axis='index', how='any', subset=['last', 'email'])

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
6     NA  Missing                     NA  Missing

În continuare o să ne uităm peste acele valori create de noi care practic și ele reprezintă tot missing values (acele valori 'NA' și 'Missing'). Putem să ne ocupăm de aceste date fie în momentul în care creem data frame-ul (cum am creat mai sus, utilizând linia de cod 'peoples_df = pd.DataFrame(peoples)'), iar dacă datele sunt citite dintr-un fișier csv, atunci procedeul o să fie puțin diferit. Pentru început o să ne uităm la cazul în care creem un data frame din acel dicționar. Pentru a putea face asta o să ne folosim de metoda 'replace()' la care o să îi atribuim valorile ce dorim să le schimbăm, cu ce anume dorim să le schimbăm

In [26]:
peoples_df

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
3  Chris  Schafer                   None       36
4    NaN      NaN                    NaN     None
5   None      NaN      Anonymus@mail.com     None
6     NA  Missing                     NA  Missing

In [27]:
peoples_df.replace('NA', np.nan, inplace=True)

In [28]:
peoples_df

first     last                  email      age
0  Corey  Schafer  CoreySchafer@mail.com       33
1   Jane      Doe      JaneDoe@gmail.com       55
2   John      Doe      JohnDoe@gmail.com       63
3  Chris  Schafer                   None       36
4    NaN      NaN                    NaN     None
5   None      NaN      Anonymus@mail.com     None
6    NaN  Missing                    NaN  Missing

Mai rămâne să modificăm doar pentru valoarea de 'Missing', să o setăm la o valorea de 'np.nan'

In [29]:
peoples_df.replace('Missing', np.nan, inplace=True)

In [30]:
peoples_df

first     last                  email   age
0  Corey  Schafer  CoreySchafer@mail.com    33
1   Jane      Doe      JaneDoe@gmail.com    55
2   John      Doe      JohnDoe@gmail.com    63
3  Chris  Schafer                   None    36
4    NaN      NaN                    NaN  None
5   None      NaN      Anonymus@mail.com  None
6    NaN      NaN                    NaN   NaN

În acest moment, după ce am modificat acele valori customatizate de missing values cu valori de tipul 'NaN', putem acuma să rulăm din nou metoda 'dropna()'

In [31]:
peoples_df.dropna()

first     last                  email age
0  Corey  Schafer  CoreySchafer@mail.com  33
1   Jane      Doe      JaneDoe@gmail.com  55
2   John      Doe      JohnDoe@gmail.com  63

O metodă de a verifica dacă anumite date se clasifică ca fiind date de tipul 'NaN', atunci în acest caz putem utiliza metoda `isna()`. Această metodă o să returneze o mască pentru data frame-ul nostru prin care se va returna 'True' pentru valorile ce sunt considerate ca și 'NaN', respectiv 'False' pentru cele care nu sunt considerate.

In [32]:
peoples_df.isna()

first   last  email    age
0  False  False  False  False
1  False  False  False  False
2  False  False  False  False
3  False  False   True  False
4   True   True   True   True
5   True   True  False   True
6   True   True   True   True

În cazurile în care lucrăm cu valori numerice există situația în care poate dorim să înlocuim aceste valori de tipul 'NaN' cu anumite valori specifice. Pentru a putea face acest lucru putem să utilizăm metoda `fillna()`. În cadrul acestei metode o să trecem valoarea cu care dorim să fie înlocuite toate valorile de tipul 'NaN'

In [33]:
peoples_df.fillna('MISSING')

first     last                  email      age
0    Corey  Schafer  CoreySchafer@mail.com       33
1     Jane      Doe      JaneDoe@gmail.com       55
2     John      Doe      JohnDoe@gmail.com       63
3    Chris  Schafer                MISSING       36
4  MISSING  MISSING                MISSING  MISSING
5  MISSING  MISSING      Anonymus@mail.com  MISSING
6  MISSING  MISSING                MISSING  MISSING

După rularea acelui cod putem observa că toate valorile ce erau considerate ca fiind 'NaN' au fost înlocuite cu 'MISSING', adică argumentul care a fost trecut pentru metoda 'fillna()'.

În continuare o să ne uităm peste un procedeu foarte des utilizat, și anume partea de modificare a tipului de date (casting). Îm cadrul acestui data frame avem o coloană care poartă denumirea de 'age'. Să presupunem că dorim să calculăm media de vârstă a persoanelor din cadrul acestui data frame. În momentul în care afișăm datele din acest data frame, valorile din coloana 'age' par a fi numere, dar dacă accesăm tipul de date o să observăm că defapt acestea sunt stringuri. Pentru a vedea tipul de date stocat în fiecare coloană, putem utiliza atributul `dtypes`

In [58]:
# changing back to default missing values
peopels_df.replace('MISSING', np.nan, inplace=True)
peopels_df.replace('Missing', np.nan, inplace=True)
peoples_df

first     last                  email   age
0  Corey  Schafer  CoreySchafer@mail.com    33
1   Jane      Doe      JaneDoe@gmail.com    55
2   John      Doe      JohnDoe@gmail.com    63
3  Chris  Schafer                   None    36
4    NaN      NaN                    NaN  None
5   None      NaN      Anonymus@mail.com  None
6    NaN      NaN                    NaN   NaN

In [59]:
peopels_df.dtypes

first    object
last     object
email    object
age      object
dtype: object

Outputul acestui cod ne afișează că în cadrul coloanei 'age' avem un tip de date denumit 'object'. Când vedem acest tip de date în cadrul unui data frame înseamnă cel mai probabil că avem parte de string ca și tip de date. Din moment ce această colană conține string-uri, pentru a afla media de vârstă nu o să funcționeze momentan. Să încercăm totuși să utilizăm metoda 'mean()' pentru coloana respectivă pentru a vedea ce eroare rezultă

In [60]:
peoples_df['age'].mean()

TypeError: can only concatenate str (not "int") to str

Eroarea este una de tipul 'TypeError' și ne spune următorul lucru: 'can only concatenate str (not "int") to str'. Nu este tocmai cea mai ușor de înțeles descriere pentru o eroare, dar practic ne spune că acea coloană este de tipul string și metoda are nevoie de date de tip integer pentru a putea funcționa corespunzător. În continuare o să modificăm tipul acelei coloane din date de tip string în integer.

Pentru cazurile în care în interiorul acelor coloane există date de tip 'NaN' atunci trebuie să transformă datele în format de tip float, deoarece valoare 'NaN' este defapt un float în spate. 

In [61]:
type(np.nan)

float

Dacă încercăm să modificăm datele din string în integer dintr-o coloană care are și valori 'NaN' atunci o să returneze o eroare deaorece nu poate converti acele valori în float. Pentru a putea converti o coloană în alt tip de date o să utilizăm metoda `astype()`. Acestei metode o să îi oferim ca și argument tipul de date în ce dorim să transformăm acea coloană. Pentru a funcționa corect, trebuie să creem o nouă coloană (sau să suprapunem coloana veche) la care să îi atribuim tot aceleași date din coloana respectivă, dar sub altă formă (alt tip de date). Pentru început o să încercăm să modificăm acea coloană în integer

In [62]:
peopels_df['age'] = peopels_df['age'].astype(int)

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'

Când încercăm să convertim o coloană din string care conține și valori de tipul 'NaN' la un tip de date numeric avem două variante. Putem să înlocuim acele valori de missing values cu un număr (precum 0 ) utilizând metoda 'fillna()' sau putem să convertim acea coloană într-o coloană de tipul float și în consecință o să ne modifice acele valori de tipul 'NaN' automat în valori de tip float (adică în 0.0). Acuma o să modificăm acea coloană în float

In [65]:
peoples_df['age'] = peopels_df['age'].astype(float)

In [66]:
peoples_df.dtypes

first     object
last      object
email     object
age      float64
dtype: object

După ce am modificat tipul de date din acea coloană, acuma putem observa că în cadrul coloanei 'age' avem date de tipul float. În acest moment putem să utilizăm metoda 'mean()' pentru a avea valoarea media de vârstă

!! Un lucru important de ținut minte este faptul că atunci când se modifică o coloană în date de tipul float, datele care sunt prezente ca și 'NaN' o să rămână la fel. Dacă era să modificăm acele date în valorea 0, atunci calculele pentru media de vărstă ar fi fost incorecte deoarece s-ar fi luat în calcul și acele valori de 0. Prin urmare, depinzând de caz, se va alege dacă să se înlocuiască acele date cu un anumit număr sau dacă să rămână tot valori de tipul 'NaN'. În cazul de față, cel mai bine este să rămână valori de tipul 'NaN' deoarece acestea nu o să fie luate în calcul pentru metoda 'mean()'

In [67]:
peopels_df

first     last                  email   age
0  Corey  Schafer  CoreySchafer@mail.com    33
1   Jane      Doe      JaneDoe@gmail.com    55
2   John      Doe      JohnDoe@gmail.com    63
3  Chris  Schafer                   None    36
4    NaN      NaN                    NaN  None
5   None      NaN      Anonymus@mail.com  None
6     NA      NaN                     NA   NaN

In [69]:
peoples_df.mean(numeric_only=True)

age    46.75
dtype: float64

În cazul în care avem un data frame care conține doar valori numerice, atunci putem să utilizăm metoda 'astype()' pentru întreg data frame-ul respectiv

In [ ]:
peoples_df.astype()

Din moment ce ne-am uitat peste aceste concepte, acuma o să le punem în practică pentru data set-ul ce conține datele despre studiul făcut de stack overflow. După cum spuneam anterior, dacă avem date de tipul missing data în cadrul unui fișier csv este mult mai ușor de a trata aceste date în momentul în care ne luăm datele într-un data frame.

In [70]:
# reading the data from csv and seting the index column
df = pd.read_csv('../data/survey_results_public.csv', index_col='Respondent')
schema_df = pd.read_csv('../data/survey_results_schema.csv', index_col='Column')

In [71]:
# setting the maximum display for rows and column to 85
pd.set_option('display.max_columns', 85)
pd.set_option('display.max_rows', 85)

Înainte de a învăța cum anume să tratăm datele de tipul missing date, metoda de sus era cea utilizată pentru a citi datele dintr-un fișier csv și a le salva în cadrul unui data frame.
Dacă dorim să tratăm datele de tip missing data în momentul în care citim un fișier csv, atunci putem să utilizăm un argument de tip listă de valori care o să conțină valori ce dorim să fie trate ca și missing values. Pentru asta o să ne creem o listă separată de valori pe care să le considerăm ca și missing values

In [73]:
na_vals = ['NA', 'Missing']

Argumentul ce trebuie utilizat pentru aceste date ce dorim să le tratăm ca și missing values poartă denumirea de `na_values`. Acestui argument o să îi oferim ca și valoare lista creată mai sus

In [74]:
df = pd.read_csv('../data/survey_results_public.csv', index_col='Respondent', na_values=na_vals)

În continuare o să ne uităm peste anumite cazuri interesante în momentul în care dorim să convertim anumite tipuri de date. Pentru acest studiu am hotărât că o să dorim să calculăm media de experiență în programare a persoanelor care au participat la acest studiu. Să vedem acuma de ce ar fi dificil acest lucru de calculat cu aceste date. Pentru a ajunge la soluția finală o să trecem prin mai multe concepte ce le-am învățat în cadrul acestui tutorial. Coloana unde sunt acele date poartă denumirea de 'YearsCode'. Să aruncăm o privire peste aceste date

In [76]:
df['YearsCode'].head(10)

Respondent
1       4
2     NaN
3       3
4       3
5      16
6      13
7       6
8       8
9      12
10     12
Name: YearsCode, dtype: object

Primele date din cadrul acesui Series nu arată să fie o problemă în momentul în care dorim să facem transformarea. Putem încerca să convertim datele din acea coloană în date de tipul float.

In [77]:
df['YearsCode'] = df['YearsCode'].astype(float)

ValueError: could not convert string to float: 'Less than 1 year'

În momentul în care dorim să convertim coloana în date de tip float primim o eroare. Eroarea respectivă ne spune că nu reușește să transforme string-ul 'Less than 1 year' în float, ceea ce este de așteptat să se întâmple așa. Din acestă eroare ne dăm seama că există alte date decât integer și 'NaN' în cadrul acestei coloane. Persoanele care au participat la studiu au avut posibilitatea de a selecta string-ul 'Less than 1 year' ca și variantă de răspuns. Pentru a vedea exact ce tip de date conține această coloană, atunci putem să utilizăm metoda `unique()` asupra unui obiect de tip Series și o să returneze toate valorile unice din cadrul acelei coloane. 

In [78]:
df['YearsCode'].unique()

array(['4', nan, '3', '16', '13', '6', '8', '12', '2', '5', '17', '10',
       '14', '35', '7', 'Less than 1 year', '30', '9', '26', '40', '19',
       '15', '20', '28', '25', '1', '22', '11', '33', '50', '41', '18',
       '34', '24', '23', '42', '27', '21', '36', '32', '39', '38', '31',
       '37', 'More than 50 years', '29', '44', '45', '48', '46', '43',
       '47', '49'], dtype=object)

În cadrul outputului de mai sus putem observa faptul că ne sunt afișate toate valorile unice din acel Series. Printre acele valori există multe numere diferite, dar putem observa și faptul că există și valori de tip string, precum 'Less than 1 year' sau 'More than 50 years'. Pentru a nu ne încurca acele valori, o să le înlocuim. Pentru 'Less than 1 year' o să înlocuim acele valori cu 0, iar pentru 'More than 50 years' o să înlocuim valorile respective cu 51. Aceste înlocuiri ne-ar putea afecta rezultatul final depinzând de cât de multe persoane au răspuns utilizând acele variante de răspuns

In [88]:
df['YearsCode'].value_counts().loc['Less than 1 year']

1367

In [87]:
df['YearsCode'].value_counts().loc['More than 50 years']

126

Putem observa faptul că 126 de persoane au răspuns cu opțiunea de 'More than 50 years', care nu reprezintă un procentaj așa de mare ca să fie o diferență mare în rezultate dacă înlocuim acea valore cu 51

In [90]:
df['YearsCode'].replace('Less than 1 year', 0, inplace=True)

In [92]:
df['YearsCode'].replace('More than 50 years', 51, inplace=True)

Putem verifica dacă aceste modificări s-au realizat uitându-ne din nou la valorile unice din cadrul acelei coloane.

In [93]:
df['YearsCode'].unique()

array(['4', nan, '3', '16', '13', '6', '8', '12', '2', '5', '17', '10',
       '14', '35', '7', 0, '30', '9', '26', '40', '19', '15', '20', '28',
       '25', '1', '22', '11', '33', '50', '41', '18', '34', '24', '23',
       '42', '27', '21', '36', '32', '39', '38', '31', '37', '29', '44',
       '45', '48', '46', '43', '47', '49'], dtype=object)

În acest moment se pare că avem valori doar de tipul integer și 'NaN', prin urmare le putem converti la un float (le convertim la float deoarece dacă dorim să le comvertim la int, atunci o să primim o eroare din cauza valorilor 'NaN')

In [94]:
df['YearsCode'] = df['YearsCode'].astype(float)

După ce am modifcat tipul datelor din coloană din string în float, acuma putem utiliza metoda 'mean()' pentru a afla media de experiență în programare a persoanelor care au participat la acest studiu

In [95]:
df['YearsCode'].mean()

11.589040005458392

## Recapitulare

În cadrul acestui tutorial am învățat următoarele lucruri:

    1. Cum să ștergem dintr-un data frame rânduri ce conțin date de tipul 'NaN'

        df.dropna()

    2. Cum să utilizăm argumente pentru metoda 'dropna()'

        df.dropna(axis='index')

            # default atributul axis este setat la valorea 'index', ceea ce înseamnă că o să șteargă tot rândul unde găsește o valoare de tipul 'NaN'
        
        df.dropna(axis='columns')

            # când atributul axis este setat la 'columns', atunci o să șteargă coloana unde găsește o valoare de tipul 'NaN'

        df.dropna(axis='index', how='any')

            # atributul 'how' reprezintă criteriile după care să se șteargă un rând sau o coloană. Default este setat la 'any', iar atunci când este setat așa, nu contează câte coloane sunt pe un anumit rând, dacă găsește o coloană cu valorea 'NaN' o să șteargă tot rândul

        df.dropna(axis='index', how='all')

            # când atributul este setat la 'all', atunci pentru a șterge un rând, fiecare valorea din fiecare coloană de pe un anumit rând trebuie să fie de tipul 'NaN'.

    3. Cum să utilizăm argumentul 'subset' pentru a șterge valori pe baza uneia sau a multor coloane

        df.dropna(axis='index', how='all', subset=['email'])

        df.dropna(axis='index', how='all', subset=['email', 'first_name'])

            # atributul 'how' utilizat împreună cu cel de 'subset' își are rostul doar când se trec mai mult de o coloană în cadrul liste pentru argumentul 'subset'

    4. Cum putem să modifică anumite valori dintr-un data set în valori de tipul NaN

        df.replace('Missing', np.nan, inplace=True)

    5. Cum putem verifica ce date dintr-un data frame sunt considerate ca fiind 'NaN'

        df.isna()

            # returnează o mască a data frame-ului cu valori de tipul 'False' și 'True'. Cele cu True sunt luate în considerare ca fiind 'NaN', iar cele cu False nu sunt

    6. Cum putem înlocui datele de tip 'NaN' dintr-un data frame cu o anumită valoare

        df.fillna('MISSING', inplace=True)

    7. Cum să verificăm tipul de date din fiecare coloană

        df.dtypes

    8. Cum să convertim dintr-un tip de date în alt tip

        df['age'] = df['age'].astype(int)

            # dacă coloana 'age' are valori de tipul 'NaN' atunci nu putem converti în integer, o să returneze o eroare. Pentru aceste cazuri trebuie să convertim în float

        df['age'] = df['age'].astype(float)
    
    9. Cum putem trata valori configurate ca fiind 'NaN' în momentul în care citim date dintr-un csv

        na_vals = ['NA', 'Missing']

        df = pd.read_csv('../data/survey_results_public.csv', index_col='Respondent', na_values=na_vals)


    10. Cum extragem valorile unice dintr-o coloană 

        df['YearsCode'].unique()
        
    